v2/yarpcerror/errors.go

```diff
@@ -59,71 +80,70 @@ func FromError(err error) *Status {
 //
 // This is always false if the error is nil.
 func IsStatus(err error) bool {
-	_, ok := err.(*Status)
+	_, ok := err.(*status)
 	return ok
 }
 
-// Status represents a YARPC error.
-type Status struct {
+// status represents a YARPC error.
+type status struct {
 	code    Code
 	name    string
 	message string
+	details interface{}
 }
 
-// WithName returns a new Status with the given name.
+// StatusOption provides options that may be called to the constructor.
+type StatusOption struct{ apply func(*status) }
+
+// WithName returns a new status with the given name.
 //
 // This should be used for user-defined errors.
-//
-// The name must only contain lowercase letters from a-z and dashes (-), and
-// cannot start or end in a dash. If the name is something else, an error with
-// code CodeInternal will be returned.
-//
-// Deprecated: Use only error codes to represent the type of the error.
-func (s *Status) WithName(name string) *Status {
-	// TODO: We plan to add a WithDetails method to add semantic metadata to
-	// Statuses soon.
-	if s == nil {
-		return nil
-	}
-	if err := validateName(name); err != nil {
-		return err.(*Status)
-	}
-	return &Status{
-		code:    s.code,
-		name:    name,
-		message: s.message,
-	}
+func WithName(name string) StatusOption {
+	return StatusOption{func(s *status) { s.name = name }}
+}
+
+// WithDetails adds semantic metadata to a Status.
+func WithDetails(details interface{}) StatusOption {
```

Comment:
Thoughts on just calling this `WithApplicationError`?

Even though we borrowed this from gRPC, I'm not a huge fan of the term `Details`.